nmtwizard/preprocess/operators/noise.py

```diff
@@ -206,6 +226,22 @@ def _apply_word_noise(self, tokens):
                 new_tokens.append(token)
         return new_tokens
 
+    def _apply_bilingual_noise(self, src_tokens, tgt_tokens):
+        if (
+            self._final_punct_add_prob > 0
+            and random.random() <= self._final_punct_add_prob
+            and all(
+                punct not in src_tokens[-1].surface for punct in self._final_punct_list
+            )
+            and all(
+                punct not in tgt_tokens[-1].surface for punct in self._final_punct_list
+            )
+        ):
+            punct_token = pyonmttok.Token(random.choice(self._final_punct_list))
+            punct_token.join_left = True
```

Comment:
This is not true for all punctuations and languages. For example !?:; in French should have a leading space.

The user should be able to configure a leading space for one side only (source or target). Maybe you can accept tuples in the punctuation list to configure different source and target values, e.g. `[".", (";", " ;"), (":", " :")]`. If there is a leading space, you should remove it and not set `join_left` when creating the new token.

What do you think?